devices/src/legacy/i8042.rs

```diff
@@ -6,47 +6,314 @@
+
+/// i8042 status register bits
+const SB_OUT_DATA_AVAIL: u8 = 1 << 0; // Data available at port 0x60
+const SB_I8042_CMD_DATA: u8 = 1 << 3; // i8042 expecting command parameter at port 0x60
+const SB_KBD_ENABLED: u8 = 1 << 4; // 1 = kbd enabled, 0 = kbd locked
+
+/// i8042 control register bits
+const CB_KBD_INT: u8 = 1 << 0; // kbd interrupt enabled
+const CB_POST_OK: u8 = 1 << 2; // POST ok (should always be 1)
+
+/// Key scan codes
+const KEY_CTRL: u16 = 0x0014;
+const KEY_ALT: u16 = 0x0011;
+const KEY_DEL: u16 = 0xE071;
+
+/// Internal i8042 buffer size, in bytes
+const BUF_SIZE: usize = 16;
 
 /// A i8042 PS/2 controller that emulates just enough to shutdown the machine.
 pub struct I8042Device {
+    /// CPU reset eventfd. We will set this event when the guest issues CMD_RESET_CPU.
     reset_evt: EventFd,
+
+    /// Keyboard interrupt event (IRQ 1).
+    kbd_interrupt_evt: EventFd,
+
+    /// The i8042 status register.
+    status: u8,
+
+    /// The i8042 control register.
+    control: u8,
+
+    /// The i8042 output port.
+    outp: u8,
+
+    /// The last command sent to port 0x64.
+    cmd: u8,
+
+    /// The internal i8042 data buffer.
+    buf: [u8; BUF_SIZE],
+    bhead: Wrapping<usize>,
+    btail: Wrapping<usize>,
 }
 
 impl I8042Device {
-    /// Constructs a i8042 device that will signal the given event when the guest requests it.
-    pub fn new(reset_evt: EventFd) -> I8042Device {
-        I8042Device { reset_evt }
+    /// Constructs an i8042 device that will signal the given event when the guest requests it.
+    pub fn new(reset_evt: EventFd, kbd_interrupt_evt: EventFd) -> I8042Device {
+        I8042Device {
+            reset_evt,
+            kbd_interrupt_evt,
+            control: CB_POST_OK | CB_KBD_INT,
+            cmd: 0,
+            outp: 0,
+            status: SB_KBD_ENABLED,
+            buf: [0; BUF_SIZE],
+            bhead: Wrapping(0),
+            btail: Wrapping(0),
+        }
+    }
+
+    /// Returns a clone of the CPU reset event fd
+    pub fn get_reset_evt_clone(&self) -> Result<EventFd> {
+        self.reset_evt.try_clone().map_err(Error::CloneCpuResetEvt)
+    }
+
+    pub fn trigger_kbd_interrupt(&self) -> Result<()> {
+        if (self.control & CB_KBD_INT) == 0 {
+            return Err(Error::KbdInterruptDisabled);
+        }
+        self.kbd_interrupt_evt
+            .write(1)
+            .map_err(Error::KbdInterruptFailure)
+    }
+
+    pub fn trigger_key(&mut self, key: u16) -> Result<()> {
+        if key & 0xff00 != 0 {
+            // Check if there is enough room in the buffer, before pushing an extended (2-byte) key.
+            if BUF_SIZE - self.buf_len() < 2 {
+                return Err(Error::InternalBufferFull);
+            }
+            self.push_byte((key >> 8) as u8)?;
+        }
+        self.push_byte((key & 0xff) as u8)?;
+
+        match self.trigger_kbd_interrupt() {
+            Ok(_) | Err(Error::KbdInterruptDisabled) => Ok(()),
```

Comment:
I think we should report an error instead of returning `Ok(())` when `trigger_kbd_interrupt` returns `Error::KbdInterruptDisabled`, because this should propagate as such when we try to to send the CTRL-ALT-DEL, which may not work if the KbdInterrupt is disabled.

Reply:
I considered that. But the fact is that, from the device model perspective, this is not an actual error. I.e. even if the guest has (temporarily) disabled the keyboard IRQ, the key presses will still end up in the i8042 buffer and will be available for when the guest decides to re-enable the IRQ. So, it's not the device that is disabled, only the notifications it's posting.

Besides, the guest OS might decide that, for whatever reason, they want to use polling instead of IRQs.

I will concede, however, that this is may conceal some unforeseen issue, so I added a log warning.